bucket_map/src/bucket.rs

```diff
@@ -528,20 +528,21 @@ impl<'b, T: Clone + Copy + PartialEq + std::fmt::Debug + 'static> Bucket<T> {
             // fail early if the data bucket we need doesn't exist - we don't want the index entry partially allocated
             return Err(BucketMapError::DataNoSpace((best_fit_bucket, 0)));
         }
+
         let max_search = self.index.max_search();
         let (elem, elem_ix) = Self::find_index_entry_mut(&self.index, key, self.random)?;
-        let elem = if let Some(elem) = elem {
-            elem
-        } else {
-            let is_resizing = false;
```

Comment:
previously we allocated the index entry here every time. On a resize of databucket, we would have gone through line 534 (`elem`) because we would have already found an allocated entry with `key`. Note that the first allocation would have been `Zero`, which is probably not valid. Now, we search and don't find it the first time through, we figure out we need to resize the data bucket, then the second time through (after resizing data bucket), we don't find it again. The allocation in the index entry occurs once we know we can put the data in the new data bucket.

Reply:
yeah. it makes sense to delay the delay the `elem` allocation after `data` allocation. 
Thanks for the explanation.